### PR TITLE
Additional provider features added and fixed some bugs

### PR DESCRIPTION
**Pull request by edevenport**
##### ISSUE TYPE
- Feature Pull Request
- Bugfix Pull Request
##### COMPONENT NAME

cloud/profitbricks/profitbricks.py
cloud/profitbricks/profitbricks_volume.py
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
###### Features
- Added support for SSH keys.
- Added support for image passwords.
- Added support for SSD disk type.
- Added support for CPU family selection.
- Updated in-code examples and documentation.
###### Bug Fixes
- Adjusted server create so that IP address is returned in response.
- Restructured remove server method(s) to handle change status properly, gracefully handle missing servers, and improve overall performance.
- Prevent duplicate server names from being provisioned so removals can be handled appropriately.
- Fixed a bug in the count increment being a string rather than an integer.
- Fixed issue with create_volume returning invalid response.
- Fixed type bug in volume instance_ids for volume removal and improved volume management.
- Fixed type bug in instance_ids for proper server removal and moved boot volume creation into composite server build request.
- General clean up.

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @edevenport. To the current maintainers, @baldwinSPC please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by baldwinSPC** [comment]
shipit

**Comment by gregdek** [comment]
Thanks again to @edevenport for this PR, and thanks @baldwinSPC for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by resmo** [comment]
Thanks @edevenport for your effort!

It seems, there are new features, bug fixes and even a new module in this PR. 

Mixing features and bug fixes prevents us from back-porting fixes to stable branches. Sometimes it can make sense if the bug was fixed by a refactored code. However in such cases a separate PR to the stable branches with a minimal patch is required e.g. stable-2.1. 

Please make a separate PR for the new modules.

Thanks for your understanding.

needs_revision

**Comment by edevenport** [comment]
@resmo Sorry about that. I inadvertently committed the new module (and a few other minor changes) to the same branch. I reverted the recent changes and have moved them to a new branch. I'll submit a new PR for that branch shortly.

**Comment by gregdek** [comment]
Thanks @edevenport for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by edevenport** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @edevenport. To the current maintainers, @baldwinSPC please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by baldwinSPC** [comment]
shipit

**Comment by gregdek** [comment]
Thanks again to @edevenport for this PR, and thanks @baldwinSPC for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/profitbricks/profitbricks.py [hunk 1]
```diff
@@ -561,12 +582,16 @@ def main():
             image=dict(),
             cores=dict(default=2),
             ram=dict(default=2048),
+            cpu_family=dict(default='AMD_OPTERON'),
```
Comment (resmo) [inline cloud/profitbricks/profitbricks.py]: please add the choice list e.g.:

``` python
cpu_family=dict(default='AMD_OPTERON', choices=[ 'AMD_OPTERON', 'INTEL_XEON']), 
```

File: cloud/profitbricks/profitbricks_volume.py [hunk 2]
```diff
@@ -308,19 +324,50 @@ def delete_volume(module, profitbricks):
...
         argument_spec=dict(
             datacenter=dict(),
+            server=dict(),
             name=dict(),
             size=dict(default=10),
             bus=dict(default='VIRTIO'),
             image=dict(),
+            image_password=dict(default=None),
+            ssh_keys=dict(type='list', default=[]),
             disk_type=dict(default='HDD'),
```
Comment (resmo) [inline cloud/profitbricks/profitbricks_volume.py]: please add choices list here as well:

``` python
 disk_type=dict(default='HDD', choices=['HDD', 'SSD']),
```

File: cloud/profitbricks/profitbricks.py [hunk 3]
```diff
@@ -561,12 +582,16 @@ def main():
...
             volume_size=dict(default=10),
+            disk_type=dict(default='HDD'),
```
Comment (resmo) [inline cloud/profitbricks/profitbricks.py]: same here:

``` python
 disk_type=dict(default='HDD', choices=['HDD', 'SSD']),
```

File: cloud/profitbricks/profitbricks_volume.py [hunk 4]
```diff
@@ -308,19 +324,50 @@ def delete_volume(module, profitbricks):
...
+        except Exception as e:
+            module.fail_json(msg='failed to attach volume: %s' % str(e))
+
 def main():
     module = AnsibleModule(
         argument_spec=dict(
             datacenter=dict(),
+            server=dict(),
             name=dict(),
             size=dict(default=10),
```
Comment (resmo) [inline cloud/profitbricks/profitbricks_volume.py]: probably set here also a `type='int',`?

File: cloud/profitbricks/profitbricks.py [hunk 5]
```diff
@@ -561,12 +582,16 @@ def main():
...
+            image_password=dict(default=None),
+            ssh_keys=dict(type='list', default=[]),
             bus=dict(default='VIRTIO'),
             lan=dict(default=1),
```
Comment (resmo) [inline cloud/profitbricks/profitbricks.py]: `type='int',`?

File: cloud/profitbricks/profitbricks.py [hunk 6]
```diff
@@ -561,12 +582,16 @@ def main():
...
-            count=dict(default=1),
```
Comment (resmo) [inline cloud/profitbricks/profitbricks.py]: `type='int',`?

File: cloud/profitbricks/profitbricks.py [hunk 7]
```diff
@@ -561,12 +582,16 @@ def main():
             image=dict(),
             cores=dict(default=2),
```
Comment (resmo) [inline cloud/profitbricks/profitbricks.py]: `type='int',`?

File: cloud/profitbricks/profitbricks.py [hunk 8]
```diff
@@ -561,12 +582,16 @@ def main():
...
             ram=dict(default=2048),
```
Comment (resmo) [inline cloud/profitbricks/profitbricks.py]: `type='int',`?

File: cloud/profitbricks/profitbricks.py (hunk 1 above)
Comment (edevenport) [reply cloud/profitbricks/profitbricks.py]: The choice list has been added.

File: cloud/profitbricks/profitbricks_volume.py (hunk 2 above)
Comment (edevenport) [reply cloud/profitbricks/profitbricks_volume.py]: The choice list has been added.

File: cloud/profitbricks/profitbricks.py (hunk 3 above)
Comment (edevenport) [reply cloud/profitbricks/profitbricks.py]: The choice list has been added.

File: cloud/profitbricks/profitbricks_volume.py (hunk 4 above)
Comment (edevenport) [reply cloud/profitbricks/profitbricks_volume.py]: Done.

File: cloud/profitbricks/profitbricks.py (hunk 5 above)
Comment (edevenport) [reply cloud/profitbricks/profitbricks.py]: Done.

File: cloud/profitbricks/profitbricks.py (hunk 7 above)
Comment (edevenport) [reply cloud/profitbricks/profitbricks.py]: Done.

File: cloud/profitbricks/profitbricks.py (hunk 8 above)
Comment (edevenport) [reply cloud/profitbricks/profitbricks.py]: And done.

**Comment by edevenport** [comment]
@resmo The changes you have suggested have been made. Thank you for the feedback.

File: cloud/profitbricks/profitbricks.py [hunk 9]
```diff
@@ -549,24 +549,50 @@ def startstop_machine(module, profitbricks, state):
...
-            volume_size=dict(default=10),
+            cores=dict(type='int', default=2),
+            ram=dict(type='int', default=2048),
+            cpu_family=dict(choices=['AMD_OPTERON', 'INTEL_XEON'],
+                            default='AMD_OPTERON'),
+            volume_size=dict(type='int', default=10),
+            disk_type=dict(choices=['HDD', 'SSD'], default='HDD'),
+            image_password=dict(default=None),
+            ssh_keys=dict(type='list', default=[]),
             bus=dict(default='VIRTIO'),
```
Comment (resmo) [inline cloud/profitbricks/profitbricks.py]: could you please also add a choice list for bus?

``` python
bus=dict(default='VIRTIO', choices=['IDE', 'VIRTIO']),
```

File: cloud/profitbricks/profitbricks_volume.py [hunk 10]
```diff
@@ -308,19 +328,52 @@ def delete_volume(module, profitbricks):
...
+
 def main():
     module = AnsibleModule(
         argument_spec=dict(
             datacenter=dict(),
+            server=dict(),
             name=dict(),
-            size=dict(default=10),
+            size=dict(type='int', default=10),
             bus=dict(default='VIRTIO'),
```
Comment (resmo) [inline cloud/profitbricks/profitbricks_volume.py]: could you please also add a choice list for bus?

``` python
bus=dict(default='VIRTIO', choices=['IDE', 'VIRTIO']),
```

**Comment by edevenport** [comment]
@resmo The parameter choices for the bus have been set.

**Comment by resmo** [comment]
@edevenport Thanks!

**Pull request merged**